typeinfo: find useful system header files by finding the maximum fixed point of compilable header files

MaskRay opened the pull request:
The trial and errors process is quite slow, it would be better to cache the list of header files somewhere. Do you where is appropriate for storing the list?

Tested on Arch Linux x86_64, extra/gdb-7.12.1-3 (Python 3)

Hi Zach!

MaskRay committed:
typeinfo: find useful system header files by finding the maximum fixed point of compilable header files.

zachriggle commented:
OMG ❤️

zachriggle commented:
There's not really a persistence mechanism for Pwndbg yet.  We haven't needed inter-session storage so far.

MaskRay commented:
Do you have a plan to add the persistence layer? I'll change the commit accordingly.

zachriggle commented:
Unfortunately, there's none currently.  I'm open to suggestions :)

disconnect3d commented:
@MaskRay Are you going to work on that too?

Here are some options for persistence directory:
* `~/.cache/pwndbg`
* `~/.local/share/pwndbg`
* `~/.config/pwndbg`
* `~/.pwndbg`

Probably would be good to check for an env var first, e.g.:
```python
persistence_path = os.environ.get('PWNDBG_PERSISTENCE_DIR', default_persistence_dir)
```

It might be good to handle it through `memoize`, e.g. adding a decorator called `forever` or `persistent`.
Another option could be an explicit `save`/`load`.

MaskRay committed:
Add pwndbg.storage and four caching API file_cache_{,with_signature_}for_{text,pickle}

Consider an imaginary file pwndbg/foo.py

@pwndbg.storage.file_cache_for_text('{0}.txt')
def ft(x):
    print(x)
    return x

The first call of ft('a') will save a file ~/.cache/pwndbg/foo/ft/a.txt .
The basename is generated from '{0}.txt'.format(x)
A later call of ft('a') will load the text from the cache.

@pwndbg.storage.file_cache_with_signature_for_text('{0}')
def fts(x):
    print(x)
    return x

The caller should provide an extra keyword argument (signature) to fts.
The first call of fts('a', signature='sig') will save a file ~/.cache/pwndbg/foo/fts/a .
A call of fts('a', signature='sig') will load the cache.
However, a call of fts('a', signature='different_sig') will recalculate as the signature is different.

MaskRay commented:
Refined and rebased on https://github.com/pwndbg/pwndbg/pull/268

disconnect3d commented:
Can we finally make it into pwndbg? If we are not really having a proper persistance mechanism, maybe we should skip it for this PR?
@zachriggle @MaskRay

disconnect3d commented:
Hmm played a bit with this and here are some thoughts:
- this wouldn't be usable without cache
- maybe it would be better to add include paths and to have all includes and remove those that are still missing? (collisions might be problematic)
- maybe we should specify some 'standard include paths' for different OS/architectures?

ps: this needs to be rebased

disconnect3d commented:
Closing as there is no reply :(
